File: ForexConnectAPIRobot/samples/java/SimpleSamples/TableManagerSamples/CreateELS/src/TableListener.java
package createels;

import com.fxcore2.IO2GTableListener;
import com.fxcore2.O2GRow;
import com.fxcore2.O2GTableStatus;
import com.fxcore2.O2GOrderRow;

public class TableListener implements IO2GTableListener {
    
    String mChangeOrderID = null;
    public String getOrderID() {
        return mChangeOrderID;
    }
    
    // Implementation of IO2GTableListener interface public method onAdded
     public void onAdded(String rowID, O2GRow rowData) { 
        O2GOrderRow orderRow = (O2GOrderRow)(rowData);
        System.out.println("Order created. Order ID is " + rowID);
        if ((orderRow.getContingencyType() == 3) && (orderRow.getType().trim().equals("LE"))) {
            mChangeOrderID = orderRow.getOrderID();
        }
    }

    // Implementation of IO2GTableListener interface public method onChanged
    public void onChanged(String rowID, O2GRow rowData) {
        System.out.println("Order changed. Order ID is " + rowID);
    }
    
    // Implementation of IO2GTableListener interface public method onDeleted
    public void onDeleted(String rowID, O2GRow rowData) {
        System.out.println("Order deleted. Order ID is " + rowID);
    }

    // Implementation of IO2GTableListener interface public method onStatus
    public void onStatusChanged(O2GTableStatus status) {
        switch(status) {
            case REFRESHED:
                System.out.println("\nYour refresh request was successfull");
                break;
            case REFRESHING:
            case INITIAL:
            case FAILED:
               break;
        }
    }
}

File: ForexConnectAPI-1.1.2-Linux-i686/samples/cpp/SimpleSamples/TableManagerSamples/NetStopLimit/source/tablelistener.cpp
#include "tablelistener.h"

TableListener::TableListener(void)
{
    mEvent = uni::CreateEvent(0, FALSE, FALSE, 0);
    mRef = 1;
    ordersState.limit = 0;
    ordersState.stop = 0;
}

TableListener::~TableListener(void)
{
    uni::CloseHandle(mEvent);
}

long TableListener::addRef()
{
    return InterlockedIncrement(&mRef);
}

long TableListener::release()
{
    InterlockedDecrement(&mRef);
    if(mRef == 0)
        delete this;
    return mRef;
}

HANDLE TableListener::getEvent()
{
    return mEvent;
}

void TableListener::onStatusChanged(O2GTableStatus status)
{
    switch(status)
    {
    case ::Refreshed:
        uni::SetEvent(mEvent);
        break;
    case ::Refreshing:
    case ::Initial:
    case ::Failed:
        break;
    }
}

void TableListener::onAdded(const char *rowID, IO2GRow *row)
{
    IO2GOrderRow *orderRow = static_cast<IO2GOrderRow *>(row);
    if ( strncmp(orderRow->getType(), O2G2::Orders::LimitEntry, 3) == 0 )
    {
        printf("added LimitEntry order. id: %s\n", orderRow->getOrderID() );
        mLimitOrderID = orderRow->getOrderID();
        ordersState.limit = 1;
        uni::SetEvent(mEvent);
        return;
    }
    if ( strncmp(orderRow->getType(), O2G2::Orders::StopEntry, 3) == 0 )
    {
        printf("added StopEntry order. id: %s\n", orderRow->getOrderID() );
        mStopOrderID = orderRow->getOrderID();
        ordersState.stop = 1;
        uni::SetEvent(mEvent);
        return;
    }
}

void TableListener::onChanged(const char *rowID, IO2GRow *row)
{
    IO2GOrderRow *orderRow = static_cast<IO2GOrderRow *>(row);
    if ( strncmp(orderRow->getType(), O2G2::Orders::LimitEntry, 3) == 0 )
    {
        printf("changed LimitEntry order. id: %s\n", orderRow->getOrderID() );
        mLimitOrderID = orderRow->getOrderID();
        ordersState.limit = 1;
        uni::SetEvent(mEvent);
        return;
    }
    if ( strncmp(orderRow->getType(), O2G2::Orders::StopEntry, 3) == 0 )
    {
        printf("changed StopEntry order. id: %s\n", orderRow->getOrderID() );
        mStopOrderID = orderRow->getOrderID();
        ordersState.stop = 1;
        uni::SetEvent(mEvent);
        return;
    }
}

void TableListener::onDeleted(const char *rowID, IO2GRow *row)
{
    IO2GOrderRow *orderRow = static_cast<IO2GOrderRow *>(row);
    if ( strncmp(orderRow->getType(), O2G2::Orders::LimitEntry, 3) == 0 )
    {
        printf("delete LimitEntry order. id: %s\n", orderRow->getOrderID() );
        mLimitOrderID = orderRow->getOrderID();
        ordersState.limit = 1;
        uni::SetEvent(mEvent);
        return;
    }
    if ( strncmp(orderRow->getType(), O2G2::Orders::StopEntry, 3) == 0 )
    {
        printf("delete StopEntry order. id: %s\n", orderRow->getOrderID() );
        mStopOrderID = orderRow->getOrderID();
        ordersState.stop = 1;
        uni::SetEvent(mEvent);
        return;
    }
}

std::string TableListener::getOrderID( const char *orderType ) const
{
    std::string res = "";
    if( ordersState.limit && ( strncmp(orderType, O2G2::Orders::LimitEntry, 3) == 0 ) )
    {
        res = mLimitOrderID;
    }
    if( ordersState.stop && ( strncmp(orderType, O2G2::Orders::StopEntry, 3) == 0 ) )
    {
        res = mStopOrderID;
    }
    return res;
}

